File: src/sorting/Sort.java
package sorting;

import java.util.Arrays;

public class Sort {

    public static void main(String[] args) {
        int[] numbers = {1, 2, 7, 3, 6, 5, 4, 9, 8};
        print(numbers);
        System.out.println();
        int[] selectionSortNumbers = selectionSort(numbers);

        numbers = new int[]{1, 2, 7, 3, 6, 5, 4, 9, 8};
        int[] insertionSortNumbers = insertionSort(numbers);

        numbers = new int[]{1, 2, 7, 3, 6, 5, 4, 9, 8};
        int[] bubbleSortNumbers = bubbleSort(numbers);

        numbers = new int[]{1, 2, 7, 3, 6, 5, 4, 9, 8};
        int[] mergeSortNumbers = mergeSort(numbers);

        numbers = new int[]{1, 2, 7, 3, 6, 5, 4, 9, 8};
        int[] quickSortNumbers = quickSort(numbers, 0, numbers.length - 1);

        numbers = new int[]{1, 2, 7, 3, 6, 5, 4, 9, 8};
        int[] heapSortNumbers = heapSort(numbers);

        print(selectionSortNumbers);
        print(insertionSortNumbers);
        print(bubbleSortNumbers);
        print(mergeSortNumbers);
        print(quickSortNumbers);
        print(heapSortNumbers);
    }

    private static void print(int[] numbers) {
        for (int i = 0; i < numbers.length; i++) {
            System.out.print(numbers[i] + " ");
        }
        System.out.println();
    }

    private static int[] selectionSort(int[] numbers) {
        // 시간복잡도 O(n^2)
        // 공간복잡도 O(1)

        for (int i = 0; i < numbers.length; i++) {
            int minIndex = i;

            for (int j = i; j < numbers.length; j++) {
                if (numbers[minIndex] > numbers[j]) {
                    minIndex = j;
                }
            }

            int temp = numbers[minIndex];
            numbers[minIndex] = numbers[i];
            numbers[i] = temp;
        }

        return numbers;
    }

    private static int[] insertionSort(int[] numbers) {
        // 시간복잡도 O(n^2) 이미 정렬되어 있는 경우는 O(n)
        // 공간복잡도 O(1)

        for (int i = 1; i < numbers.length; i++) {
            for (int j = 1; j < i; j++) {
                if (numbers[j - 1] < numbers[i] && numbers[i] <= numbers[j]) {
                    int temp = numbers[i];
                    numbers[i] = numbers[j];
                    numbers[j] = temp;
                    break;
                }
            }
        }

        return numbers;
    }

    private static int[] bubbleSort(int[] numbers) {
        // 시간복잡도 O(n^2)
        // 공간복잡도 O(1)

        for (int i = 0; i < numbers.length; i++) {
            for (int j = 1; j < numbers.length - i; j++) {
                if (numbers[j - 1] > numbers[j]) {
                    int temp = numbers[j - 1];
                    numbers[j - 1] = numbers[j];
                    numbers[j] = temp;
                }
            }
        }

        return numbers;
    }

    private static int[] mergeSort(int[] numbers) {
        // 시간복잡도 O(NlogN)
        // 공간복잡도 O(N)

        if (numbers.length < 2) {
            return numbers;
        }

        int mid = numbers.length / 2;
        int[] low = mergeSort(Arrays.copyOfRange(numbers, 0, mid));
        int[] high = mergeSort(Arrays.copyOfRange(numbers, mid, numbers.length));

        return sort(low, high);
    }

    private static int[] sort(int[] low, int[] high) {
        int[] result = new int[low.length + high.length];

        int lowPoint = 0;
        int highPoint = 0;

        for (int i = 0; i < result.length; i++) {
            if (lowPoint == low.length) {
                result[i] = high[highPoint];
                highPoint++;
                continue;
            }

            if (highPoint == high.length) {
                result[i] = low[lowPoint];
                lowPoint++;
                continue;
            }

            if (low[lowPoint] < high[highPoint]) {
                result[i] = low[lowPoint];
                lowPoint++;
            } else {
                result[i] = high[highPoint];
                highPoint++;
            }
        }

        return result;
    }

    private static int[] quickSort(int[] numbers, int left, int right) {
        // 시간복잡도 O(NlogN) 최악의 경우(pivot값이 편향적일 경우) O(N^2)
        // 공간복잡도 O(logN)

        int i, j, pivot, tmp;

        if (left < right) {
            i = left;
            j = right;
            pivot = numbers[(left + right) / 2];

            while (i < j) {
                while (numbers[j] > pivot) j--;
                while (numbers[i] < pivot) i++;

                tmp = numbers[j];
                numbers[j] = numbers[i];
                numbers[i] = tmp;
            }


            quickSort(numbers, left, i - 1);
            quickSort(numbers, i + 1, right);
        }

        return numbers;
    }

    private static int[] heapSort(int[] numbers) {
        // 시간 복잡도 O(NlogN)
        // 공간 복잡도 O(1)

        for (int i = numbers.length / 2 - 1; i >= 0; i--) {
            heapify(numbers, numbers.length, i);
        }

        for (int i = numbers.length - 1; i >= 0; i--) {
            int temp = numbers[0];
            numbers[0] = numbers[i];
            numbers[i] = temp;

            heapify(numbers, i,0);
        }

        return numbers;
    }

    private static void heapify(int[] numbers, int size, int node) {
        int parent = node;
        int leftChild = node * 2 + 1;
        int rightChild = node * 2 + 2;

        if (leftChild < size && numbers[leftChild] > numbers[parent]) {
            parent = leftChild;
        }

        if (rightChild < size && numbers[rightChild] > numbers[parent]) {
            parent = rightChild;
        }

        if (parent != node) {
            int temp = numbers[node];
            numbers[node] = numbers[parent];
            numbers[parent] = temp;

            heapify(numbers, size, parent);
        }
    }
}

File: python/book_python_algorithm_interview/ch17/sort.py
# 버블 정렬(Bubble Sort)
# n번의 라운드로 이뤄져 있으며, 각 라운드마다 배열의 아이템을 한 번씩 모두 살펴본다.
# 연달아 있는 아이템 2개의 순서가 잘못되어 있는 것을 발견하면, 두 아이템을 맞바꾼다.
# 시간 복잡도 : O(n^2)
# (스왑 여부를 확인해 더 최적화할 수 있지만, 그럼에도 다른 정렬 알고리즘과는 여전히 성능 차이가 있음)
def bubblesort(A):
    for i in range(1, len(A)):
        for j in range(0, len(A) - 1):
            if A[j] > A[j + 1]:
                A[j], A[j + 1] = A[j + 1], A[j]


# 병합 정렬(Merge Sort)
# 분할 정복(Divide and Conquer)을 이용한 알고리즘
# 더 이상 쪼갤 수 없을 때까지 계속해서 분할한 후, 분할이 끝나면 정렬하면서 정복해 나간다.
# 퀵 정렬보다는 느리지만 일정한 실행 속도 뿐만 아니라 안정 정렬이라는 점에서 상용 라이브러리에 많이 쓰임
# 시간 복잡도 : 최선, 최악 모두 O(n log n)
def mergesort(A):
    if len(A) < 2:
        return A

    mid = len(A) // 2
    low = mergesort(A[:mid])
    high = mergesort(A[mid:])

    result = []
    l = h = 0
    while l < len(low) and h < len(high):
        if low[l] < high[h]:
            result.append(low[l])
            l += 1
        else:
            result.append(high[h])
            h += 1

    result += low[l:]
    result += high[h:]
    return result

# 퀵 정렬(Quick Sort)
# 병합 정렬과 마찬가지로 분할 정복 알고리즘을 사용
# 피벗(Pivot)이라는 개념을 통해 피벗보다 작으면 왼쪽, 크면 오른쪽과 같은 방식으로 파티셔닝하면서 쪼개 나간다.
# 여러 가지 변형과 개선 버전이 있는데 여기서는 로무토 파티션 계획을 살펴본다.
# 로무토 파티션이란 항상 맨 오른쪽의 피벗을 택하는 단순한 방법
# (피봇을 어떤 값으로 선정하느냐에 따라 성능 차이가 난다. 예를 들어, 이미 정렬된 배열이 들어온 경우)
# 시간 복잡도 : O(n log n), 최악 O(n^2)


def quicksort(A, lo, hi):
    def partition(lo, hi):
        pivot = A[hi]
        left = lo
        for right in range(lo, hi):
            if A[right] < pivot:
                A[left], A[right] = A[right], A[left]
                left += 1
        A[left], A[hi] = A[hi], A[left]
        return left

    if lo < hi:
        pivot = partition(lo, hi)
        quicksort(A, lo, pivot - 1)
        quicksort(A, pivot + 1, hi)
